core/src/main/java/org/apache/iceberg/TableScanContext.java

```diff
@@ -353,17 +352,29 @@ TableScanContext planWith(ExecutorService executor) {
         toSnapshotId,
         executor,
         fromSnapshotInclusive,
-        metricsReporters);
+        metricsReporter);
   }
 
-  Collection<MetricsReporter> metricsReporters() {
-    return metricsReporters;
+  MetricsReporter metricsReporter() {
+    return metricsReporter;
   }
 
   TableScanContext reportWith(MetricsReporter reporter) {
-    ImmutableList.Builder<MetricsReporter> builder = ImmutableList.builder();
-    builder.addAll(metricsReporters);
-    builder.add(reporter);
+    CompositeMetricsReporter compositeReporter;
+    if (metricsReporter instanceof CompositeMetricsReporter) {
+      compositeReporter = (CompositeMetricsReporter) metricsReporter;
+    } else {
+      compositeReporter = new CompositeMetricsReporter();
```

Comment:
Rather than creating a reporter and calling register, the pattern we use most often is to have a factory method. That factory method can detect duplicates and return a single reporter, or can return a composite reporter.

```java
  MetricsReporter r1 = MetricsReporters.combine(LoggingReporter.INSTANCE, LoggingReporter.INSTANCE); // produces LoggingReporter.INSTANCE
  MetricsReporter c1 = MetricsReporters.combine(r1, r2); // produces CompositeMetricsRerporter
  MetricsReporter c2 = MetricsReporters.combine(c1, r3); // produces CompositeMetricsRerporter
```

Reply:
thanks for the suggestion, I like that